### PR TITLE
Add BinaryReader Span-based APIs

### PR DESCRIPTION
**Pull request by kevingosse**
Contributes to https://github.com/dotnet/corefx/issues/22428

Depends on https://github.com/dotnet/coreclr/pull/13616, https://github.com/dotnet/corert/pull/4396

### CODE REVIEW

**Comment by KooKiz** [comment]
Is `src/System.IO/tests/BinaryReader/BinaryReaderTests.cs` the only available tests for the BinaryReader? If so, I may need to add non-regression tests for `ReadChars(int count)` and `Read(char[] buffer, int index, int count)` since they both depend on `InternalReadChars`.

**Comment by danmosemsft** [comment]
@KooKiz looks like the tests are all there, except indirect coverage in src\System.IO\tests\BinaryWriter\**. More would be welcome.

**Comment by pjanotti** [comment]
@dotnet-bot test this please

**Comment by kevingosse** [comment]
@dotnet-bot test this please

**Comment by kevingosse** [comment]
@dotnet-bot test this please

**Comment by kevingosse** [comment]
@dotnet-bot test this please

File: src/System.IO/tests/BinaryReader/BinaryReaderTests.cs [hunk 1]
```diff
@@ -74,6 +75,8 @@ private void ValidateDisposedExceptions(BinaryReader binaryReader)
             Assert.Throws<ObjectDisposedException>(() => binaryReader.Read());
             Assert.Throws<ObjectDisposedException>(() => binaryReader.Read(byteBuffer, 0, 1));
             Assert.Throws<ObjectDisposedException>(() => binaryReader.Read(charBuffer, 0, 1));
+            Assert.Throws<ObjectDisposedException>(() => binaryReader.Read(new Span<byte>(byteBuffer)));
+            Assert.Throws<ObjectDisposedException>(() => binaryReader.Read(new Span<char>(charBuffer)));
```
Comment (stephentoub) [inline src/System.IO/tests/BinaryReader/BinaryReaderTests.cs]: Tests for the new APIs will need to be separated out into their own file conditionally included only for netcoreapp.

Comment (kevingosse) [reply src/System.IO/tests/BinaryReader/BinaryReaderTests.cs]: Done.

**Comment by kevingosse** [comment]
@dotnet-bot test this please

**Comment by kevingosse** [comment]
@dotnet-bot test this please

**Comment by kevingosse** [comment]
@dotnet-bot test this please

**Comment by kevingosse** [comment]
@dotnet-bot test this please

**Comment by kevingosse** [comment]
@dotnet-bot test this please

**Pull request merged**